src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs

```diff
@@ -18159,5 +18391,399 @@ namespace System
+
+class Program
+{
+    static void Main()
+    {
+        Helper1<S>.Test1(null);
+        System.Console.Write(""_"");
+        Helper1<S>.Test1(new Helper1<S>());
+    }
+}
+";
+
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, options: TestOptions.ReleaseExe);
+            var verifier = CompileAndVerify(
+                comp,
+                expectedOutput: ExecutionConditionUtil.IsMonoOrCoreClr ? "_M1" : null,
+                verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).VerifyDiagnostics();
+
+            verifier.VerifyIL("Helper1<T>.Test1(Helper1<T>)",
+@"
+{
+  // Code size       11 (0xb)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  brfalse.s  IL_000a
+  IL_0003:  ldarg.0
+  IL_0004:  call       ""T Helper1<T>.M1()""
+  IL_0009:  pop
+  IL_000a:  ret
+}
+");
+        }
+
+        [Fact]
+        public void DynamicAccess_01()
+        {
+            var src = @"
+class Helper1<T>
+    where T : struct, allows ref struct
+{
+    static void Test1(dynamic h1)
+    {
+        h1.M1<T>();
+    }
+}
+
+class Helper2
+{
+    static void Test2(dynamic h2)
+    {
+        h2.M2<S>();
+    }
+}
+
+ref struct S
+{
+}
+";
+
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+            comp.VerifyDiagnostics(
+                // (7,15): error CS0306: The type 'T' may not be used as a type argument
+                //         h1.M1<T>();
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "T").WithArguments("T").WithLocation(7, 15),
+                // (15,15): error CS0306: The type 'S' may not be used as a type argument
+                //         h2.M2<S>();
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "S").WithArguments("S").WithLocation(15, 15)
+                );
+        }
+
+        [Fact]
+        public void DynamicAccess_02()
+        {
+            var src = @"
+class Helper1<T>
+    where T : struct, allows ref struct
+{
+    static void Test1(dynamic h1, T x)
+    {
+        h1.M1(x);
+    }
+}
+
+class Helper2
+{
+    static void Test2(dynamic h2, S y)
+    {
+        h2.M2(y);
+    }
+}
+
+ref struct S
+{
+}
+";
+
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+            comp.VerifyDiagnostics(
+                // (7,15): error CS1978: Cannot use an expression of type 'T' as an argument to a dynamically dispatched operation.
+                //         h1.M1(x);
+                Diagnostic(ErrorCode.ERR_BadDynamicMethodArg, "x").WithArguments("T").WithLocation(7, 15),
+                // (15,15): error CS1978: Cannot use an expression of type 'S' as an argument to a dynamically dispatched operation.
+                //         h2.M2(y);
+                Diagnostic(ErrorCode.ERR_BadDynamicMethodArg, "y").WithArguments("S").WithLocation(15, 15)
+                );
+        }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

Are we testing a ref struct receiver and dynamic argument?
```csharp
class Helper1<T>
    where T : I1, allows ref struct
{
    static void Test1(T x, dynamic d)
    {
        x.M(d);
    }
}
interface I1
{
    void M<T>(T t);
}
```

Reply:
> Are we testing a ref struct receiver and dynamic argument?

Adding `DynamicAccess_03` for completeness.